File: 20. Valid Parentheses.py
def isValid(s):
    """
    :type s: str
    :rtype: bool
    """
    buf = []
    opbuf = {'(': ')', '[': ']', '{':'}'}
    
    for i in s:
        if i in opbuf:
            buf.append(opbuf[i])
        elif i in opbuf.values():
            if (not buf) or buf[-1] != i:
                return False
            else:
                buf.pop()
    if not buf:
        return True
    else:
        return False

s = "()"
print(isValid(s))

File: 20. Valid Parentheses.go
package main

import "fmt"

func isValid(s string) bool {
	buf := make([]byte, 0)
	for _, c := range s {
		switch c {
		case '(':
			buf = append(buf, ')')
		case '[':
			buf = append(buf, ']')
		case '{':
			buf = append(buf, '}')
		case ')', ']', '}':
			if len(buf) == 0 {
				return false
			}
			if rune(buf[len(buf)-1]) == rune(c) {
				buf = buf[:len(buf)-1]
			} else {
				return false
			}
		}
	}

	if len(buf) == 0 {
		return true
	} else {
		return false
	}
}

func main() {

	s := "()"
	fmt.Println(isValid(s))
}
